modules/images/webp-uploads/generation.php

```diff
@@ -0,0 +1,236 @@
+ * @param string $image_src     The 'src' of the image.
+ * @param array  $image_meta    The image meta data as returned by 'wp_get_attachment_metadata()'.
+ * @param int    $attachment_id Image attachment ID or 0.
+ * @return array Srcset sources array.
+ */
+function webp_uploads_generate_image_srcset( $sources, $size_array, $image_src, $image_meta, $attachment_id ) {
+	$image_src_parts = wp_parse_url( $image_src );
+	$image_host      = $image_src_parts['host'];
+	if ( ! empty( $image_src_parts['port'] ) ) {
+		$image_host .= ':' . $image_src_parts['port'];
+	}
+
+	// Do nothing if images are loaded from a different origin.
+	if ( $image_host !== $_SERVER['HTTP_HOST'] ) {
+		return $sources;
+	}
+
+	// Do nothing if the image attachment ID is not supplied.
+	if ( empty( $attachment_id ) ) {
+		return $sources;
+	}
+
+	require_once ABSPATH . '/wp-admin/includes/image.php';
+
+	// Return early if there are no missing sizes.
+	$missing_sizes = wp_get_missing_image_subsizes( $attachment_id );
+	if ( empty( $missing_sizes ) ) {
+		return $sources;
+	}
+
+	// Retrieve the uploads sub-directory from the full size image.
+	$dirname = _wp_get_attachment_relative_path( $image_meta['file'] );
+	if ( $dirname ) {
+		$dirname = trailingslashit( $dirname );
+	}
+
+	$upload_dir    = wp_get_upload_dir();
+	$image_baseurl = trailingslashit( $upload_dir['baseurl'] ) . $dirname;
+
+	/*
+	 * If currently on HTTPS, prefer HTTPS URLs when we know they're supported by the domain
+	 * (which is to say, when they share the domain name of the current request).
+	 */
+	if ( is_ssl() && 'https' !== substr( $image_baseurl, 0, 5 ) && parse_url( $image_baseurl, PHP_URL_HOST ) === $_SERVER['HTTP_HOST'] ) {
+		$image_baseurl = set_url_scheme( $image_baseurl, 'https' );
+	}
+
+	$ext  = pathinfo( $image_meta['file'], PATHINFO_EXTENSION );
+	$name = wp_basename( $image_meta['file'], ".$ext" );
+	foreach ( $missing_sizes as $size_name => $size_details ) {
+		$dimensions = wp_constrain_dimensions(
+			$size_array[0],
+			$size_array[1],
+			$size_details['width'],
+			$size_details['height']
+		);
+
+		$sources[ $dimensions[0] ] = array(
+			'url'        => sprintf( '%s%s-%dx%d.%s', $image_baseurl, $name, $dimensions[0], $dimensions[1], $ext ),
+			'descriptor' => 'w',
+			'value'      => $dimensions[0],
+		);
+	}
+
+	return $sources;
+}
+add_filter( 'wp_calculate_image_srcset', 'webp_uploads_generate_image_srcset', 10, 5 );
+
+/**
+ * Updates the srcset attribute for images in the content. Adds missing sizes to the image.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $img           An <img> tag.
+ * @param string $context       The context where this is function is being used.
+ * @param int    $attachment_id The ID of the attachment being modified.
+ * @return string The updated img tag.
+ */
+function webp_uploads_update_img_tag_srcset( $img, $context, $attachment_id ) {
+	list( $src, $width, $height ) = wp_get_attachment_image_src( $attachment_id, 'large', false );
```

Comment:
Do we want to review that the returned value from `wp_get_attachment_image_src` is not `false` or that the values from `$src`, `$width`, and `$height` are not `null`?